File: src/simulate.ts
/**
 * @module simulate
 * Advanced transaction simulation guardrail.
 * Natively executes `eth_call` to trace and simulate an agent's intended transaction
 * against the current blockchain state. Prevents MEV sandwich attacks, honeypot drains,
 * and wasted gas by failing closed if the simulation reverts.
 */

import { createPublicClient, http, type PublicClient, type Hex } from "viem";
import { mainnet } from "viem/chains";

export type SimulateGuardrailResult = {
  decision: "ALLOW" | "BLOCK";
  reason: string;
  allowExecute: boolean;
};

export type SimulateTxInput = {
  rpc: string;
  account: string;
  to: string;
  data?: string;
  value?: bigint;
  /** @internal inject mock client for tests */
  __client?: Pick<PublicClient, "call">;
};

/**
 * Simulates a transaction against the latest block.
 * If the simulation reverts (e.g., honeypot tax, slippage tripped, insufficient gas),
 * this guardrail returns BLOCK.
 * Agents MUST check this before executing live transactions.
 */
export async function simulateTx(input: SimulateTxInput): Promise<SimulateGuardrailResult> {
  const { rpc, account, to, data, value } = input;

  if (!/^0x[a-fA-F0-9]{40}$/.test(account)) {
    return {
      decision: "BLOCK",
      reason: `invalid account address ${account} — BLOCK`,
      allowExecute: false,
    };
  }
  if (!/^0x[a-fA-F0-9]{40}$/.test(to)) {
    return { decision: "BLOCK", reason: `invalid to address ${to} — BLOCK`, allowExecute: false };
  }
  if (data !== undefined && !/^0x[a-fA-F0-9]*$/.test(data)) {
    return { decision: "BLOCK", reason: `invalid data hex payload — BLOCK`, allowExecute: false };
  }
  if (value !== undefined && (typeof value !== "bigint" || value < 0n)) {
    return { decision: "BLOCK", reason: `invalid value — BLOCK`, allowExecute: false };
  }
  if (typeof rpc !== "string" || rpc.trim() === "") {
    return { decision: "BLOCK", reason: "missing rpc — BLOCK (fail closed)", allowExecute: false };
  }

  const client =
    input.__client ??
    createPublicClient({
      chain: mainnet,
      transport: http(rpc),
    });

  try {
    // `client.call` executes `eth_call`. If the smart contract reverts, this throws.
    await client.call({
      account: account as `0x${string}`,
      to: to as `0x${string}`,
      data: (data || "0x") as Hex,
      value: value ?? 0n,
    });
  } catch (err) {
    const msg = err instanceof Error ? err.message : String(err);
    // Viem gives extremely detailed revert messages, which is perfect for AI agents
    // to digest and adjust their parameters.
    return {
      decision: "BLOCK",
      reason: `Simulation reverted! The transaction will fail on-chain. Do NOT execute. Revert reason: ${msg}`,
      allowExecute: false,
    };
  }

  return {
    decision: "ALLOW",
    reason: `transaction simulation succeeded`,
    allowExecute: true,
  };
}

File: src/simulate.rs
//! Pre-flight execution simulation (`eth_call`).
//!
//! Revert → BLOCK. Check-then-act advisory: state can shift between
//! simulation and broadcast (frontrun, slot churn). Bind outcomes with
//! deadlines, slippage limits, and private mempools.

use crate::addressbook::is_valid_eth_address;
use crate::rpc::EvmRpcClient;
use crate::types::GuardrailResult;

#[derive(Debug, Clone)]
/// `account`: simulation sender. `to`: target. `data`: calldata (`0x` if none).
/// `value`: native wei to send; `None` simulates a 0-value call (legacy).
/// Payable flows MUST set `value` — simulating them as 0-value would
/// approve a transaction whose real execution path was never tested.
pub struct SimulateTxInput<'a> {
    pub account: &'a str,
    pub to: &'a str,
    pub data: Option<&'a str>,
    pub value: Option<u128>,
}

/// Simulate via `eth_call(from, to, data[, value])`. Any revert/transport
/// error → BLOCK.
pub async fn simulate_tx(client: &dyn EvmRpcClient, input: SimulateTxInput<'_>) -> GuardrailResult {
    if !is_valid_eth_address(input.account) {
        return GuardrailResult::block(format!(
            "invalid account address {} — BLOCK",
            input.account
        ));
    }
    if !is_valid_eth_address(input.to) {
        return GuardrailResult::block(format!("invalid to address {} — BLOCK", input.to));
    }

    let calldata = input.data.unwrap_or("0x");

    let simulated = match input.value {
        Some(wei) => {
            client
                .call_from_with_value(input.account, input.to, calldata, wei)
                .await
        }
        None => client.call_from(input.account, input.to, calldata).await,
    };

    match simulated {
        Ok(_) => GuardrailResult::allow("transaction simulation succeeded"),
        Err(e) => GuardrailResult::block(format!(
            "Simulation reverted! The transaction will fail on-chain. Do NOT execute. Revert reason: {}",
            e
        )),
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::mock::MockRpcClient;
    use std::sync::Arc;

    #[tokio::test]
    async fn test_simulation_succeeds() {
        let mock = MockRpcClient {
            call_handler: Some(Arc::new(move |_, _| Ok("0x".to_string()))),
            ..Default::default()
        };

        let res = simulate_tx(
            &mock,
            SimulateTxInput {
                account: "0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045",
                to: "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48",
                data: None,
                value: None,
            },
        )
        .await;

        assert!(res.allow_execute);
    }

    #[tokio::test]
    async fn test_simulation_reverts_blocks() {
        let mock = MockRpcClient {
            call_handler: Some(Arc::new(move |_, _| {
                Err("reverted with UniswapV2: K".to_string())
            })),
            ..Default::default()
        };

        let res = simulate_tx(
            &mock,
            SimulateTxInput {
                account: "0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045",
                to: "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48",
                data: None,
                value: None,
            },
        )
        .await;

        assert!(!res.allow_execute);
        assert!(res.reason.contains("Simulation reverted"));
    }

    #[tokio::test]
    async fn test_simulation_forwards_value() {
        let mock = MockRpcClient {
            call_from_value_handler: Some(Arc::new(move |_, _, _, wei| {
                assert_eq!(wei, 1_000_000_000_000_000_000);
                Ok("0x".to_string())
            })),
            ..Default::default()
        };

        let res = simulate_tx(
            &mock,
            SimulateTxInput {
                account: "0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045",
                to: "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48",
                data: None,
                value: Some(1_000_000_000_000_000_000),
            },
        )
        .await;

        assert!(res.allow_execute);
    }

    #[tokio::test]
    async fn test_simulation_value_without_handler_blocks() {
        // Value-bearing simulation with no value handler must BLOCK, never
        // silently fall back to a 0-value simulation.
        let mock = MockRpcClient {
            call_handler: Some(Arc::new(move |_, _| Ok("0x".to_string()))),
            ..Default::default()
        };

        let res = simulate_tx(
            &mock,
            SimulateTxInput {
                account: "0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045",
                to: "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48",
                data: None,
                value: Some(1),
            },
        )
        .await;

        assert!(!res.allow_execute);
        assert!(res.reason.contains("Simulation reverted"));
    }
}
